### PR TITLE
feat(authz-keycloak): add token refresh/expire time configuration

### PR DESCRIPTION
**Pull request by Belyenochi**
resolved https://github.com/apache/apisix/issues/3479

### CODE REVIEW

**Review by spacewander — COMMENTED** [review]
No feature added?

**Comment by tzssangglass** [comment]
> No feature added?

The code has been implemented, but not covered by the documentation and test cases, https://github.com/apache/apisix/blob/ec0fc2ceaf04a20b0bd0ebdaad67296a1d3f621c/apisix/plugins/authz-keycloak.lua#L316-L327

**Comment by tzssangglass** [comment]
pls add en docs too.

**Comment by tzssangglass** [comment]
pls add description of `access_token_expires_in` and `refresh_token_expires_in`

**Comment by Belyenochi** [comment]
> pls add description of `access_token_expires_in` and `refresh_token_expires_in`

I will add the test case

**Review by spacewander — COMMENTED** [review]
Need to add these fields in https://github.com/apache/apisix/blob/9b98f1d13187c58195543bd48cb8b66e33aeab27/apisix/plugins/authz-keycloak.lua#L29

**Comment by tzssangglass** [comment]
> The code has been implemented, but not covered by the documentation and test cases,

In fact, the description of these two parameters is also missing

**Comment by Belyenochi** [comment]
> > The code has been implemented, but not covered by the documentation and test cases,
> 
> In fact, the description of these two parameters is also missing

I have tried adding test cases and descriptions.

File: docs/zh/latest/plugins/authz-keycloak.md [hunk 1]
```diff
@@ -38,15 +38,19 @@ title: authz-keycloak
...
+| permissions                 | array[string] | 可选   |             |                             | 描述客户端应用所需访问的资源和权限范围的字符串。格式必须为：`RESOURCE_ID#SCOPE_ID`              |
+| timeout                     | integer       | 可选   | 3000        | [1000, ...]                 | 与身份认证服务器的 http 连接的超时时间                                                          |
+| access_token_expires_in     | integer       | 可选   | 300         | [1, ...]                            | access token的过期时间
+| access_token_expires_leeway | integer       | 可选   | 0           | [0, ...]                            | access token提前到期时间（如果设置了此值，允许在该时间段内使用相同的access token令牌来解决潜在的网络并发问题）        |
+| refresh_token_expires_in    | integer       | 可选   | 3600        | [1, ...]                            | refresh token的过期时间                                                                               |
+| refresh_token_expires_leeway| integer       | 可选   | 0           | [0, ...]                            | refresh token提前更新时间（如果设置了此值，允许在该时间段内使用相同的refresh token令牌来解决潜在的网络并发问题）                                                                   |
```
Comment (tzssangglass) [inline docs/zh/latest/plugins/authz-keycloak.md:51]:
```suggestion
| access_token_expires_in     | integer       | 可选   | 300         | [1, ...]                            | access token 的过期时间
| access_token_expires_leeway | integer       | 可选   | 0           | [0, ...]                            | access token 提前到期时间（如果设置了此值，允许在该时间段内使用相同的 access token 令牌来解决潜在的网络并发问题）        |
| refresh_token_expires_in    | integer       | 可选   | 3600        | [1, ...]                            | refresh token 的过期时间                                                                               |
| refresh_token_expires_leeway| integer       | 可选   | 0           | [0, ...]                            | refresh token 提前更新时间（如果设置了此值，允许在该时间段内使用相同的 refresh token 令牌来解决潜在的网络并发问题）                                                                   |
```

File: docs/zh/latest/plugins/authz-keycloak.md [hunk 2]
```diff
@@ -123,6 +127,7 @@ docker run -e KEYCLOAK_USER=admin -e KEYCLOAK_PASSWORD=123456 -p 8090:8080 sshni
 
 ## 后续开发
 
-- 目前 `authz-plugin` 仅支持通过定义资源名和访问权限范畴来应用 `route` 的访问策略。但是 Keycloak 官方适配的其他语言的客户端 (Java, JS) 还可以通过动态查询 Keycloak 路径以及懒加载身份资源的路径来支持路径匹配。未来版本的 `authz-plugin` 将会支持这项功能。
+- 目前 `authz-plugin` 仅支持通过定义资源名和访问权限范畴来应用 `route` 的访问策略。但是 Keycloak 官方适配的其他语言的客户端 (Java, JS) 还可以通过动态查询 Keycloak
+  路径以及懒加载身份资源的路径来支持路径匹配。未来版本的 `authz-plugin` 将会支持这项功能。
```
Comment (tzssangglass) [inline docs/zh/latest/plugins/authz-keycloak.md:131]: why change this?

Comment (Belyenochi) [reply docs/zh/latest/plugins/authz-keycloak.md:131]: Sorry I ran a format code by mistake in the IDE, I'll fix it soon

**Comment by Belyenochi** [comment]
@tzssangglass The fix has been completed, thanks for Review

**Comment by tzssangglass** [comment]
hi @Belyenochi , thank you for contributing, pls fix Doc Lint check,just a little problem that needs to be fixed.

File: apisix/plugins/authz-keycloak.lua [hunk 3]
```diff
@@ -61,8 +61,13 @@ local schema = {
...
         keepalive_timeout = {type = "integer", minimum = 1000, default = 60000},
-        keepalive_pool = {type = "integer", minimum = 1, default = 5}
-    },
+        keepalive_pool = {type = "integer", minimum = 1, default = 5},
+        access_token_expires_in = {type = "integer", minimum = 1, default = 300},
+        access_token_expires_leeway = {type = "integer", minimum = 0, default = 0},
+        refresh_token_expires_in = {type = "integer", minimum = 1, default = 3600},
+        refresh_token_expires_leeway = {type = "integer", minimum = 0, default = 0},
+
+},
```
Comment (spacewander) [inline apisix/plugins/authz-keycloak.lua:70]: Please fix the indent and remove blank line

File: docs/en/latest/plugins/authz-keycloak.md [hunk 4]
```diff
@@ -54,6 +54,10 @@ For more information on Keycloak, refer to [Keycloak Authorization Docs](https:/
...
 | http_method_as_scope           | boolean       | optional    | false                                         |                                                                    | Map HTTP request type to scope of same name and add to all permissions requested.                                                                           |
 | timeout                        | integer       | optional    | 3000                                          | [1000, ...]                                                        | Timeout(ms) for the http connection with the Identity Server.                                                                                               |
+| access_token_expires_in        | integer       | optional    | 300                                           | [1, ...]                                                           | The expiration time of the access token                                                                                                                                                            |
```
Comment (spacewander) [inline docs/en/latest/plugins/authz-keycloak.md:57]: Need to add time unit in the doc

File: apisix/plugins/authz-keycloak.lua [hunk 5]
```diff
@@ -61,8 +61,13 @@ local schema = {
         cache_ttl_seconds = {type = "integer", minimum = 1, default = 24 * 60 * 60},
         keepalive = {type = "boolean", default = true},
         keepalive_timeout = {type = "integer", minimum = 1000, default = 60000},
-        keepalive_pool = {type = "integer", minimum = 1, default = 5}
-    },
+        keepalive_pool = {type = "integer", minimum = 1, default = 5},
+        access_token_expires_in = {type = "integer", minimum = 1, default = 300},
```
Comment (spacewander) [inline apisix/plugins/authz-keycloak.lua:65]: We can remove the `or` logic now?
https://github.com/apache/apisix/blob/ecf08c6dd50c3ea3e7bf3c4723ada5e90bbeae13/apisix/plugins/authz-keycloak.lua#L318

**Comment by spacewander** [comment]
> hi @Belyenochi , thank you for contributing, pls fix Doc Lint check,just a little problem that needs to be fixed.

Solved in https://github.com/apache/apisix/commit/1c6b47354ac4bc4f04b21cb626eafaf0d875bafb

File: apisix/plugins/authz-keycloak.lua (hunk 3 above)
Comment (Belyenochi) [reply apisix/plugins/authz-keycloak.lua:69]: hi spacewander, thank you for reviewing and I'm not sure where the indent needs to be adjusted.

Comment (Belyenochi) [reply apisix/plugins/authz-keycloak.lua:69]: > hi @spacewander, thank you for reviewing and I'm not sure where the indent needs to be adjusted.

File: apisix/plugins/authz-keycloak.lua [hunk 6]
```diff
@@ -315,15 +319,15 @@ end
 
 -- Return access_token expires_in value (in seconds).
 local function authz_keycloak_access_token_expires_in(conf, expires_in)
-    return (expires_in or conf.access_token_expires_in or 300)
-           - 1 - (conf.access_token_expires_leeway or 0)
+    return (expires_in or conf.access_token_expires_in)
+           - 1 - (conf.access_token_expires_leeway)
```
Comment (spacewander) [inline apisix/plugins/authz-keycloak.lua:323]: We can remove the `()` now.

**Review by spacewander — DISMISSED** [review]

**Review by spacewander — APPROVED** [review]

**Comment by Belyenochi** [comment]
@tzssangglass hi tzssangglass, pls help to review and merge this PR.

**Review by tzssangglass — APPROVED** [review]

**Pull request merged**